src/neptune/api/models.py

```diff
@@ -0,0 +1,490 @@
+#
+# Copyright (c) 2024, Neptune Labs Sp. z o.o.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import annotations
+
+__all__ = (
+    "FileEntry",
+    "Field",
+    "FieldType",
+    "GitCommit",
+    "LeaderboardEntry",
+    "LeaderboardEntriesSearchResult",
+    "FieldVisitor",
+    "FloatField",
+    "IntField",
+    "BoolField",
+    "StringField",
+    "DateTimeField",
+    "FileField",
+    "FileSetField",
+    "FloatSeriesField",
+    "StringSeriesField",
+    "ImageSeriesField",
+    "StringSetField",
+    "GitRefField",
+    "ObjectStateField",
+    "NotebookRefField",
+    "ArtifactField",
+    "FieldDefinition",
+)
+
+import abc
+from dataclasses import dataclass
+from dataclasses import field as dataclass_field
+from datetime import datetime
+from enum import Enum
+from typing import (
+    Any,
+    ClassVar,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Set,
+    Type,
+    TypeVar,
+)
+
+from neptune.internal.utils.iso_dates import parse_iso_date
+from neptune.internal.utils.run_state import RunState
+
+Ret = TypeVar("Ret")
+
+
+@dataclass
+class FileEntry:
+    name: str
+    size: int
+    mtime: datetime
+    file_type: str
+
+    @classmethod
+    def from_dto(cls, file_dto: Any) -> "FileEntry":
+        return cls(name=file_dto.name, size=file_dto.size, mtime=file_dto.mtime, file_type=file_dto.fileType)
+
+
+class FieldType(Enum):
+    FLOAT = "float"
+    INT = "int"
+    BOOL = "bool"
+    STRING = "string"
+    DATETIME = "datetime"
+    FILE = "file"
+    FILE_SET = "fileSet"
+    FLOAT_SERIES = "floatSeries"
+    STRING_SERIES = "stringSeries"
+    IMAGE_SERIES = "imageSeries"
+    STRING_SET = "stringSet"
+    GIT_REF = "gitRef"
+    OBJECT_STATE = "experimentState"
+    NOTEBOOK_REF = "notebookRef"
+    ARTIFACT = "artifact"
+
+
+@dataclass
+class Field(abc.ABC):
+    path: str
+    type: FieldType = dataclass_field(init=False)
+    _registry: ClassVar[Dict[str, Type[Field]]] = {}
+
+    def __init_subclass__(cls, *args: Any, field_type: FieldType, **kwargs: Any) -> None:
+        super().__init_subclass__(*args, **kwargs)
+        cls.type = field_type
```

Comment:
`ClassVar` is rather special. It's needed to prevent `@dataclass` from processing this property. It doesn't have an impact on typing as far as I know.

Reply:
yeah its only about consistency.
Line `x: ClassVar[T]` inside class `A` definition means that `x` is attribute on the level of classes and subclasses of `A`  not their instances (even tho they still have access to this attribute).

So imo `type` is as much `ClassVar` as `_registry` but it's a minor typing detail so I have no problem with leaving it as it is now.

Reply:
Done. I misunderstood the question.